src/zenml/orchestrators/step_runner.py

```diff
@@ -197,75 +198,110 @@ def run(
-                            self.configuration.success_hook_source
+                    except BaseException as step_exception:  # noqa: E722
+                        step_failed = True
+                        retries += 1
+                        failure_hook_source = (
+                            self.configuration.failure_hook_source
                         )
-                        if success_hook_source:
-                            logger.info("Detected success hook. Running...")
+                        if failure_hook_source:
+                            logger.info("Detected failure hook. Running...")
                             self.load_and_run_hook(
-                                success_hook_source,
-                                step_exception=None,
+                                failure_hook_source,
+                                step_exception=step_exception,
                             )
-
-                        # Store and publish the output artifacts of the step function.
-                        output_data = self._validate_outputs(
-                            return_values, output_annotations
-                        )
-                        artifact_metadata_enabled = is_setting_enabled(
-                            is_enabled_on_step=step_run_info.config.enable_artifact_metadata,
-                            is_enabled_on_pipeline=step_run_info.pipeline.enable_artifact_metadata,
-                        )
-                        artifact_visualization_enabled = is_setting_enabled(
-                            is_enabled_on_step=step_run_info.config.enable_artifact_visualization,
-                            is_enabled_on_pipeline=step_run_info.pipeline.enable_artifact_visualization,
-                        )
-                        output_artifact_ids = self._store_output_artifacts(
-                            output_data=output_data,
-                            output_artifact_uris=output_artifact_uris,
-                            output_materializers=output_materializers,
-                            output_annotations=output_annotations,
-                            artifact_metadata_enabled=artifact_metadata_enabled,
-                            artifact_visualization_enabled=artifact_visualization_enabled,
+                        if retries < max_retries:
+                            logger.error(
+                                f"Step '{step_run_info.step_run_id}' failed. "
+                                f"Retrying in {delay} seconds."
+                            )
+                            time.sleep(delay)
+                            delay *= backoff
+                        else:
+                            logger.error(
+                                f"Step '{step_run_info.step_run_id}' failed after "
+                                f"{retries} retries."
+                            )
+                            raise
+                    finally:
```

Comment:
Looks like it will go into infinite retries upon success, cause you don't break or increase `retries` variable anywhere except `except` clause